data-connection/src/main/java/io/micronaut/data/connection/support/AbstractConnectionOperations.java

```diff
@@ -45,13 +46,41 @@ public abstract class AbstractConnectionOperations<C> implements ConnectionOpera
 
     protected final Logger logger = LoggerFactory.getLogger(getClass());
 
+    private final List<ConnectionListener<C>> connectionListeners;
+
+    protected AbstractConnectionOperations(List<ConnectionListener<C>> connectionListeners) {
+        this.connectionListeners = connectionListeners;
+    }
+
     /**
      * Opens a new connection.
      *
      * @param definition The connection definition
      * @return The connection
      */
-    protected abstract C openConnection(ConnectionDefinition definition);
+    protected final C openConnection(ConnectionDefinition definition) {
+        ConnectionStatus<C> connectionStatus = doOpenConnection(definition);
+        for (ConnectionListener<C> connectionListener : connectionListeners) {
```

Comment:
Bind listeners to using `registerSynchronization` to avoid rechecks if the connection is supported